mgowrapper.go

```diff
@@ -0,0 +1,213 @@
+package mgohttp
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	opentracinglog "github.com/opentracing/opentracing-go/log"
+	mgo "gopkg.in/mgo.v2"
+	bson "gopkg.in/mgo.v2/bson"
+)
+
+type tracedMgoSession struct {
+	sess *mgo.Session
+	ctx  context.Context
+}
+
+func (ts tracedMgoSession) DB(name string) MongoDatabase {
+	sp := opentracing.SpanFromContext(ts.ctx)
+	sp.SetTag("db-name", name)
+	return tracedMgoDatabase{
+		db:  ts.sess.DB(name),
+		ctx: opentracing.ContextWithSpan(ts.ctx, sp),
+	}
+}
+
+func (ts tracedMgoSession) Ping() error {
+	sp, _ := opentracing.StartSpanFromContext(ts.ctx, "ping")
+	defer sp.Finish()
+
+	return logAndReturnErr(sp, ts.sess.Ping())
+}
+
+type tracedMgoDatabase struct {
+	db  *mgo.Database
+	ctx context.Context
+}
+
+func (t tracedMgoDatabase) C(collection string) MongoCollection {
+	return tracedMgoCollection{
+		collectionName: collection,
+		collection:     t.db.C(collection),
+		ctx:            t.ctx,
+	}
+}
+
+func (t tracedMgoDatabase) Run(cmd interface{}, result interface{}) error {
+	sp, _ := opentracing.StartSpanFromContext(t.ctx, "run")
+	defer sp.Finish()
+	sp.LogKV(opentracinglog.String("cmd", fmt.Sprintf("%#v", cmd)))
+
+	return logAndReturnErr(sp, t.db.Run(cmd, result))
+}
+
+type tracedMgoCollection struct {
+	collectionName string
+	collection     *mgo.Collection
+	ctx            context.Context
+}
+
+func (tc tracedMgoCollection) Update(selector interface{}, update interface{}) error {
+	sp, _ := opentracing.StartSpanFromContext(tc.ctx, "update")
+	sp.SetTag("collection", tc.collectionName)
+	sp.LogFields(queryToFields("selector", selector)...)
+	sp.LogFields(queryToFields("update", update)...)
+	defer sp.Finish()
+
+	return logAndReturnErr(sp, tc.collection.Update(selector, update))
+}
+
+func (tc tracedMgoCollection) Insert(docs ...interface{}) (err error) {
+	sp, _ := opentracing.StartSpanFromContext(tc.ctx, "insert")
+	sp.LogFields(opentracinglog.Int("num-docs", len(docs)))
+	defer sp.Finish()
+
+	return logAndReturnErr(sp, tc.Insert(docs...))
+}
+func (tc tracedMgoCollection) Upsert(selector interface{}, update interface{}) (info *mgo.ChangeInfo, err error) {
+	sp, _ := opentracing.StartSpanFromContext(tc.ctx, "upsert")
+	sp.LogFields(queryToFields("selector", selector)...)
+	sp.LogFields(queryToFields("update", update)...)
+	defer sp.Finish()
+
+	info, err = tc.Upsert(selector, update)
+	return info, logAndReturnErr(sp, err)
+}
+
+func (tc tracedMgoCollection) Find(selector interface{}) MongoQuery {
+	sp, ctx := opentracing.StartSpanFromContext(tc.ctx, "find")
+	sp.SetTag("collection", tc.collectionName)
+
+	// NOTE: Find just starts the trace, the finishing call on the MongoQuery must
```

Comment:
Would it make sense to copy+paste this for other calls that just start traces (e.g. Limit)?

Reply:
agreed 👍 
added more comments in https://github.com/Clever/mgohttp/pull/7/commits/2615aee3870df473e1b0be59b90c6bc8ba45c1a1